gluonnlp/model/train/embedding.py

```diff
@@ -239,70 +261,246 @@ def __init__(self, token_to_idx, subword_function, embedding_size,
+        mxnet.ndarray.NDArray:
+            The embedding vector(s) of the token(s). According to numpy
+            conventions, if `tokens` is a string, returns a 1-D NDArray
+            (vector); if `tokens` is a list of strings, returns a 2-D NDArray
+            (matrix) of shape=(len(tokens), vec_len).
+        """
+        squeeze = False
+        if isinstance(tokens, _str_types):
+            tokens = [tokens]
+            squeeze = True
+
+        vecs = []
+
+        ctx = self.embedding.weight.list_ctx()[0]
+        for token in tokens:
+            if token in self.token_to_idx:
+                # Word is part of fastText model
+                word = nd.array([self.token_to_idx[token]], ctx=ctx)
+                wordmask = nd.ones_like(word)
+            else:
+                word = nd.array([0], ctx=ctx)
+                wordmask = nd.zeros_like(word)
+            subwords = self.subword_function([token])[0].expand_dims(0)
+            subwords = subwords.as_in_context(ctx)
+            if subwords.shape[1]:
+                vec = self(word, subwords, wordsmask=wordmask)
+            else:
+                # token is a special_token and subwords are not taken into account
+                assert token in self.token_to_idx
+                vec = self.embedding(word)
+
+            vecs.append(vec)
+
+        if squeeze:
+            assert len(vecs) == 1
+            return vecs[0].squeeze()
+        else:
+            return nd.concat(*vecs, dim=0)
+
+    def __call__(self, words, subwords, wordsmask=None, subwordsmask=None,
+                 words_to_unique_subwords_indices=None):
+        return super(FasttextEmbeddingModel, self).__call__(
+            words, subwords, wordsmask, subwordsmask,
+            words_to_unique_subwords_indices)
+
+    def forward(self, words, subwords, wordsmask=None, subwordsmask=None,
+                words_to_unique_subwords_indices=None):
         """Compute embedding of words in batch.
 
         Parameters
         ----------
         words : mx.nd.NDArray
             Array of token indices.
-        wordsmask : mx.nd.NDArray
-            Mask for embeddings returend by the word level embedding operator.
         subwords : mx.nd.NDArray
-            The subwords associated with the tokens in `words`.
-        subwordsmask : mx.nd.NDArray
+            The subwords associated with the tokens in `words`. If
+            words_to_unique_subwords_indices is specified may contain the
+            subwords of the unique tokens in `words` with
+            `words_to_unique_subwords_indices` containing the reverse mapping.
+        wordsmask : mx.nd.NDArray, optional
+            Mask for embeddings returend by the word level embedding operator.
+        subwordsmask : mx.nd.NDArray, optional
             A mask for the subword embeddings looked up from `subwords`.
             Applied before sum reducing the subword embeddings.
+        words_to_unique_subwords_indices : mx.nd.NDArray, optional
```

Comment:
word_indices are not unique, why?

Reply:
deduplicate

Reply:
Optionally, the block allows to input the `subwords` and `subwordsmask` only for the unique word_indices. Then `words_to_unique_subwords_indices` allows to map back from the non-unique `words` to the unique subword vectors. The idea is not to duplicate computing the vectors based on the ngrams (https://github.com/dmlc/gluon-nlp/pull/159/files/778fc7863505b269e7f62c23d40c3031397efb81#diff-dfdfe68db95e68d351b9222e5cce274cR308)

Reply:
This was previously blocked by https://github.com/apache/incubator-mxnet/issues/11314